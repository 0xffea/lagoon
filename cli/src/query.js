// @flow

import R from 'ramda';
import { config, configDefaults } from './config';
import { getApiConfig } from './config/getApiConfig';
import { fileExists, readFile } from './util/fs';
import request from './util/request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function runGQLQuery({
  cerr,
  query,
  variables,
  headers: customHeaders = {},
  pretty = false,
}:
QLQueryArgs): Object {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = R.prop('token', { ...configDefaults, ...config });
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const { hostname, port } = getApiConfig();

  const options = {
    hostname,
    path: '/graphql',
    port,
    method: 'POST',
    headers,
    body: JSON.stringify(
      {
        query,
        variables,
      },
      null,
      pretty ? 2 : 0,
    ),
    rejectUnauthorized: false,
  };

  try {
    return request(options);
  } catch (err) {
    const error = R.ifElse(
      // For socket hang ups...
      R.propEq('message', 'socket hang up'),
      // ...print a nicer error message...
      R.always('Could not connect to Lagoon API.'),
      // ...otherwise just return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, { message: error });
    process.exit(1);
  }
}

export default runGQLQuery;
