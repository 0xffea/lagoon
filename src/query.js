// @flow

import url from 'url';
import request from './util/request';

type QLQueryArgs = {
  endpoint: string,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function runGQLQuery(args: QLQueryArgs): Promise<Object> {
  const {
    endpoint,
    query,
    variables,
    headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
    },
    pretty = false,
  } = args;

  const {
    hostname,
    path,
    port,
  } = url.parse(endpoint);

  if (hostname == null) {
    throw new Error('Hostname required');
  }

  const body = JSON.stringify({
    query,
    variables,
  }, null, (pretty ? 2 : 0));

  const options = {
    hostname,
    path,
    port: port || 443,
    method: 'POST',
    headers,
    body,
    rejectUnauthorized: false,
  };

  return request(options);
}

export default runGQLQuery;
