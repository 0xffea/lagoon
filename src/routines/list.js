// @flow

import { take, put, go, chan, takeAsync, spawn } from 'js-csp'; // eslint-disable-line
import { contains } from 'ramda';
import gql from '../gql';

import {
  stdoutEffect,
  stderrEffect,
  gqlEffect,
  stateEffect,
  connectPut,
  effectHandler,
} from '../effects';

import type { EffectHandlerArgs } from '../effects';
import typeof { Channel } from 'js-csp/es/impl/channels';
import typeof RunGQLQueryFn from '../query';

type ListSitesArgs = {
  input: Channel,
  out: Channel,
};

export function* listSites(args: ListSitesArgs): Generator<*, *, *> {
}

function* end(input: Channel, out: Channel, code: number): Object {
  yield put(input, stateEffect(code));
  const state = yield take(out);

  input.close();
  out.close();

  return state;
}

type CliRoutineArgs = EffectHandlerArgs & {
  routine: Generator<*, *, *>,
};

export function* runRoutine(args: CliRoutineArgs) {
  const { routine } = args;

  go(effectHandler, [args]);
  return go(routine, [args]);
}

type ListArgs = {
  input: Channel,
  out: Channel,
  sitegroup: string,
  target: 'sites',
};

export default function* list(args: ListArgs): Generator<*, *, *> {
  const {
    input,
    out,
    sitegroup,
    target,
    runGQLQueryFn,
  } = args;

  const log = connectPut(input, stdoutEffect);
  const logErr = connectPut(input, stderrEffect);
  const gqlQuery = connectPut(input, gqlEffect);
  const getState = connectPut(input, stateEffect);

  if (!contains(target, ['sites'])) {
    yield logErr(`Unknown target '${target}'`);
    return yield take(go(end, [input, out, 1]));
  }

  const query = gql`query myQuery($sitegroup: String!) {
    siteGroupByName(name: $sitegroup) {
      gitUrl
      sites(first: 1000) {
        edges {
          node {
            siteName
            siteBranch
            siteEnvironment
            siteHost
            serverInfrastructure
            serverIdentifier
            serverNames
          }
        }
      }
    }
  }`;

  yield gqlQuery(query, { sitegroup });
  const result = yield take(out);

  yield log(result);

  yield getState(0);
  const { code, stack } = yield take(out);

  yield log(`State code: ${code}`);
  yield log(stack);

  input.close();
  return code;
}
