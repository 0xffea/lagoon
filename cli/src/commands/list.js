// @flow

/* eslint-disable no-console */

import { table } from 'table';
import { red } from 'chalk';
import {
  prepend,
  pathOr,
  prop,
  propOr,
  map,
  compose,
  // $FlowIssue: sortBy is not yet working: https://github.com/flowtype/flow-typed/blob/a9e64f62729fa4ecec82648f0a3f47afeff77574/definitions/npm/ramda_v0.x.x/flow_v0.34.x-/ramda_v0.x.x.js#L460-L462
  sortBy,
  toLower,
} from 'ramda';

import gql from '../gql';
import { runGQLQuery } from '../query';
import { exitNoConfig, exitError, exitGraphQLError } from '../exit';

import typeof { default as Yargs } from 'yargs';
import type { BaseArgs } from './index';

const name = 'list';
const description = 'Lists specific deployment information';

export async function setup(yargs: Yargs): Promise<Object> {
  return yargs
    .usage(`$0 ${name} [target] - ${description}`)
    .options({
      sitegroup: {
        demandOption: false,
        describe:
          'Overrides the currently configured sitegroup (.amazeeio.yml)',
        type: 'string',
      },
    })
    .alias('s', 'sitegroup')
    .example(
      `$0 ${name} sites`,
      'Lists all sites for the specific sitegroup configured in your .amazeeio.yml config file',
    )
    .example(
      `$0 ${name} sites -s mysitegroup`,
      'Lists all sites for a specific sitegroup (instead of using the config file)',
    ).argv;
}

type Target = 'sites';

type Args = BaseArgs & {
  sitegroup: ?string,
  target: Target,
};

export async function run(args: Args): Promise<number> {
  const { config, clog = console.log } = args;

  if (config == null) {
    return exitNoConfig(clog);
  }

  const [target] = args._.slice(1);
  const sitegroup = args.sitegroup || config.sitegroup;

  switch (target) {
    case 'sites':
      return listSites({ sitegroup });
    default:
      return exitError(
        clog,
        `Unknown target ${target} ... possible values: 'sites'`,
        1,
      );
  }
}

type MainArgs = {
  sitegroup: string,
  clog?: typeof console.log,
};

export async function listSites(args: MainArgs): Promise<number> {
  const { sitegroup, clog = console.log } = args;

  const query = gql`
    query querySites($sitegroup: String!) {
      siteGroupByName(name: $sitegroup) {
        gitUrl
        sites(first: 1000) {
          edges {
            node {
              siteName
              siteBranch
              siteEnvironment
            }
          }
        }
      }
    }
  `;

  const result = await runGQLQuery({
    query,
    variables: { sitegroup },
  });

  const { errors } = result;
  if (errors != null) {
    return exitGraphQLError(clog, errors);
  }

  const sortBySite = sortBy(compose(toLower, propOr('', 'siteName')));

  const nodes = compose(
    sortBySite,
    map(edge => prop('node', edge)),
    pathOr([], ['data', 'siteGroupByName', 'sites', 'edges']),
  )(result);

  if (nodes.length === 0) {
    clog(red(`No sites found for sitegroup '${sitegroup}'`));
    return 0;
  }

  clog(`I found following sites for sitegroup '${sitegroup}':`);

  const tableConfig = {
    columns: {
      // $FlowIssue: Flow doesn't understand numbers as keys https://github.com/facebook/flow/issues/380
      0: {
        // eslint-disable-line quote-props
        alignment: 'left',
        minWidth: 15,
      },
      // $FlowIssue: Flow doesn't understand numbers as keys https://github.com/facebook/flow/issues/380
      1: {
        // eslint-disable-line quote-props
        alignment: 'left',
        minWidth: 15,
      },
      // $FlowIssue: Flow doesn't understand numbers as keys https://github.com/facebook/flow/issues/380
      2: {
        // eslint-disable-line quote-props
        alignment: 'center',
        minWidth: 15,
      },
    },
  };

  const tableBody = map(node => {
    const inProdMarker = node.siteEnvironment === 'production' ? '\u221A' : '';
    return [node.siteName, node.siteBranch, inProdMarker];
  }, nodes);

  const tableData = prepend(['Site', 'Branch', 'Deployed?'], tableBody);

  clog(table(tableData, tableConfig));

  return 0;
}

export default {
  setup,
  name,
  description,
  run,
};
