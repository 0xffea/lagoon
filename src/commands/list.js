// @flow

/* eslint-disable no-console */

import { red } from 'chalk';
import { pathOr, prop, forEach, map, compose } from 'ramda';

import gql from '../gql';
import { runGQLQuery } from '../query';

import typeof { default as Yargs } from 'yargs';
import type { BaseArgs } from './index';

const name = 'list';
const description = 'Lists specific deployment information';

export async function setup(yargs: Yargs) {
  return yargs
    .usage(`$0 ${name} [target] - ${description}`)
    .options({
      sitegroup: {
        alias: 's',
        demand: false,
        describe: 'Overrides the currently configured sitegroup (.amazeeio.yml)',
      },
    })
    .example(`$0 ${name} sites`, 'Lists all sites for the specific sitegroup configured in your .amazeeio.yml config file')
    .example(`$0 ${name} sites -s mysitegroup`, 'Lists all sites for a specific sitegroup (instead of using the config file)')
    .argv;
}

function errorWithMessage(message: string, code?: number = 1): number {
  console.log(message);
  return code;
}

type Target = 'sites';

type Args = BaseArgs & {
  sitegroup: ?string,
  target: Target,
};

export async function run(args: Args): Promise<number> {
  const [target] = args._.slice(1);
  const sitegroup = args.sitegroup || args.config.sitegroup;

  switch (target) {
    case 'sites': return listSites({ sitegroup });
    default: return errorWithMessage(`Unknown target ${target}`, 1);
  }
}

type MainArgs = {
  sitegroup: string,
  clog?: typeof console.log,
};

export async function listSites(args: MainArgs): Promise<number> {
  const {
    sitegroup,
    clog = console.log,
  } = args;

  const query = gql`query querySites($sitegroup: String!) {
    siteGroupByName(name: $sitegroup) {
      gitUrl
      sites(first: 1000) {
        edges {
          node {
            siteName
          }
        }
      }
    }
  }`;

  const result = await runGQLQuery({
    endpoint: 'https://amazeeio-api-staging.herokuapp.com/graphql',
    query,
    variables: { sitegroup },
  });

  if (result.errors != null) {
    clog(red('Oops! Server sent us some errors:'));
    forEach(({ message }) => clog(`-> ${message}`), result.errors);
    return 1;
  }

  const nodes =
    compose(
      map((edge) => prop('node', edge)),
      pathOr([], ['data', 'siteGroupByName', 'sites', 'edges'])
    )(result);

  if (nodes.length === 0) {
    clog(red(`No sites found for sitegroup '${sitegroup}'`));
    return 0;
  }

  clog(`I found following sites for sitegroup '${sitegroup}'`);

  forEach((node) => {
    clog(`|- ${node.siteName}`);
  }, nodes);

  return 0;
}

export default {
  name,
  description,
  run,
};
