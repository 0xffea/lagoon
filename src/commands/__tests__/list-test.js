// @flow

import { runGQLQuery } from '../../query';
import { listSites } from '../list';

jest.mock('../../query');

const _mock = (mockFn: any): JestMockFn => mockFn;

const mockResponse = {
  data: {
    siteGroupByName: {
      sites: {
        edges: [
          {
            node: { siteName: 'site1' },
          },
          {
            node: { siteName: 'site2' },
          },
        ],
      },
    },
  },
};

describe('listSites', () => {
  it('should list sites as given by GraphQL', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() => Promise.resolve(mockResponse));

    const clog = jest.fn();

    const code = await listSites({
      sitegroup: 'some_sitegroup',
      clog,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should show error message if GraphQL returns errors', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() => Promise.resolve({
      errors: [{ message: 'Something, something missing parameter X' }],
    }));

    const clog = jest.fn();

    const code = await listSites({
      sitegroup: 'some_sitegroup',
      clog,
    });

    expect(code).toBe(1);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should show message for non-existing sites', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() => Promise.resolve({}));

    const clog = jest.fn();

    const code = await listSites({
      sitegroup: 'some_sitegroup',
      clog,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
