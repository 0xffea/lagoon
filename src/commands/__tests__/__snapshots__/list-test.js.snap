exports[`listSites should list sites as given by GraphQL 1`] = `
Array [
  Array [
    "I found following sites for sitegroup 'some_sitegroup'",
  ],
  Array [
    "|- site1",
  ],
  Array [
    "|- site2",
  ],
]
`;

exports[`listSites should show error message if GraphQL returns errors 1`] = `
Array [
  Array [
    "[31mOops! Server sent us some errors:[39m",
  ],
  Array [
    "-> Something, something missing parameter X",
  ],
]
`;

exports[`listSites should show message for non-existing sites 1`] = `
Array [
  Array [
    "[31mNo sites found for sitegroup 'some_sitegroup'[39m",
  ],
]
`;
